smack-experimental/src/main/java/org/jivesoftware/smackx/file_metadata/element/FileMetadataElement.java

```diff
@@ -0,0 +1,308 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.file_metadata.element;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.EqualsUtil;
+import org.jivesoftware.smack.util.HashCode;
+import org.jivesoftware.smack.util.StringUtils;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+import org.jivesoftware.smackx.hashes.HashManager;
+import org.jivesoftware.smackx.hashes.element.HashElement;
+
+/**
+ * File metadata element as defined in XEP-0446: File Metadata Element.
+ *
+ * This element is used in a generic way to provide information about files, eg. during file sharing.
+ */
+public final class FileMetadataElement implements ExtensionElement {
+
+    public static final String ELEMENT = "file";
+    public static final String NAMESPACE = "urn:xmpp:file:metadata:0";
+    public static final String ELEM_DATE = "date";
+    public static final String ELEM_DIMENSIONS = "dimensions";
+    public static final String ELEM_DESC = "desc";
+    public static final String ELEM_LENGTH = "length";
+    public static final String ELEM_MEDIA_TYPE = "media-type";
+    public static final String ELEM_NAME = "name";
+    public static final String ELEM_SIZE = "size";
+    public static final String ELEM_THUMBNAIL = "thumbnail";
+
+
+    private final Date date;
+    private final String dimensions;
+    private final Map<String, String> descriptions = new ConcurrentHashMap<>();
+    private final Map<HashManager.ALGORITHM, HashElement> hashElements = new ConcurrentHashMap<>();
+    private final Long length;
+    private final String mediaType;
+    private final String name;
+    private final Long size;
+    private final Set<ExtensionElement> otherElements = new HashSet<>();
+
+    private FileMetadataElement(Date date, String dimensions, Map<String, String> descriptions,
+                               Map<HashManager.ALGORITHM, HashElement> hashElements, Long length,
+                               String mediaType, String name, Long size,
+                               Set<ExtensionElement> otherElements) {
+        this.date = date;
+        this.dimensions = dimensions;
+        this.descriptions.putAll(descriptions);
```

Comment:
In such cases, you may want to consider to use Smack's `CollectionUtil.cloneAndSeal()` helper method, to ensure that the build class is truly immutable